File: boj/BinarySearch/BOJ_15810.py
# 백준 15810 풍선 공장
# 21.02.14

import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
sys.setrecursionlimit(10**6)

n, m = map(int, sys.stdin.readline().split())

times = list(map(int, sys.stdin.readline().split()))
ans = right = min(times) * m
left = 0
mid = (left + right) // 2

while left <= right:
    balloons = 0
    for i in times:
        balloons += mid // i
    
    if balloons >= m:
        right = mid - 1
        ans = min(ans, mid)
    else: left = mid + 1
    mid = (left + right) // 2

print(ans)

File: boj/BinarySearch/BOJ_15810.swift
//  백준 15810 풍선 공장
//  21.02.17

import Foundation
let INF: Int = 2147000000

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0], m = input[1]

var times = readLine()!.split(separator: " ").map { Int(String($0))! }
var l = 0, r = times.min()!*m
var mid = (l + r) / 2, ans = r

while l <= r {
  var cnt = 0
  for i in times {
    cnt += mid / i
  }
  
  if cnt >= m {
    ans = min(ans, mid)
    r = mid - 1
  } else {
    l = mid + 1
  }
  mid = (l + r) / 2
}

print(ans)
